extensions/spring-xd-extension-batch/src/main/java/org/springframework/batch/step/tasklet/x/AbstractProcessBuilderTasklet.java

```diff
@@ -73,25 +88,71 @@ public RepeatStatus execute(StepContribution contribution, ChunkContext chunkCon
 
 		String commandClassPath = createClassPath(this.getClass());
 
-		ProcessBuilder pb = new ProcessBuilder(command).redirectErrorStream(true);
+		File out = File.createTempFile(commandName + "-", ".out");
+		stepExecution.getExecutionContext().putString(commandName.toLowerCase() + ".system.out", out.getAbsolutePath());
+		logger.info(commandName + " system.out: " + out.getAbsolutePath());
+		File err = File.createTempFile(commandName + "-", ".err");
+		stepExecution.getExecutionContext().putString(commandName.toLowerCase() + ".system.err", err.getAbsolutePath());
+		ProcessBuilder pb = new ProcessBuilder(command).redirectOutput(out).redirectError(err);
 		Map<String, String> env = pb.environment();
 		env.put("CLASSPATH", commandClassPath);
 		String msg = commandDescription + " is being launched";
 		stepExecution.getExecutionContext().putString(commandName.toLowerCase() + ".command", commandDisplayString.trim());
-		List<String> commandLog = new ArrayList<String>();
+		List<String> commandOut = new ArrayList<String>();
+		List<String> commandErr = new ArrayList<String>();
+		Process p = null;
 		try {
-			Process p = pb.start();
-			p.waitFor();
-			exitCode = p.exitValue();
-			msg = commandDescription + " finished with exit code: " + exitCode;
-			if (exitCode == 0) {
+			p = pb.start();
+
+			while (!complete) {
+
+				try {
+					exitCode = p.exitValue();
+					complete = true;
+				} catch (IllegalThreadStateException e) {
+					if (stopped) {
+						p.destroy();
+						p = null;
+						break;
+					}
+				}
+
+				if (!complete) {
+
+					if (stoppable) {
+						JobExecution jobExecution =
+								jobExplorer.getJobExecution(stepExecution.getJobExecutionId());
+
+						if (jobExecution.isStopping()) {
+							stopped = true;
+						} else if (chunkContext.getStepContext().getStepExecution().isTerminateOnly()) {
+							stopped = true;
+						}
+					}
```

Comment:
Yes, for the Soop job that's not useful though, but for other process it might be.